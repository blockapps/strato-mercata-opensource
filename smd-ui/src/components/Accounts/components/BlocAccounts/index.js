import React, { Component } from 'react';
import {
  fetchAccounts,
  changeAccountFilter,
  fetchUserAddresses,
  fetchAccountDetail,
  resetUserAddress,
  fetchOauthAccounts,
} from '../../accounts.actions';
import mixpanelWrapper from '../../../../lib/mixpanelWrapper';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import SendTokens from '../SendTokens';
import Account from '../../../Account';
import CreateBlocUser from '../../../CreateBlocUser';
import { isOauthEnabled } from '../../../../lib/checkMode';

class BlocAccounts extends Component {

  constructor() {
    super()
    this.state = {
      selected: 0
    }
  }

  updateFilter(filter) {
    this.props.changeAccountFilter(filter);
  };

  onUserClick(user, address, index) {
    if (address.length && index === this.state.selected) {
      this.props.resetUserAddress(user);
      this.setState({ selected: null });
    } else {
      mixpanelWrapper.track('accounts_row_click');
      this.props.fetchUserAddresses(user, true, this.props.selectedChain)
    }
  }

  render() {
    const accounts = this.props.accounts;
    const filter = this.props.filter;
    const users = Object.getOwnPropertyNames(accounts);
    const rows = [];
    const selectedAddresses = [];
    const isModeOauth = isOauthEnabled();

    users.filter(user => {
      if (!filter) {
        return true;
      }
      return user
        .toLowerCase()
        .indexOf(filter) > -1
    })
      .forEach(function (user, index) {
        const addresses = Object.getOwnPropertyNames(accounts[user]);
        let userClasseName = '';
        if (this.state.selected === index && addresses.length > 0) {
          userClasseName = ' selected';
          addresses.map(address =>
            selectedAddresses.push(<Account name={user} address={address} key={address} />)
          );
        }

        rows.push(
          <div className="smd-margin-8" key={user}>
            <div className="row">
              <div className={`pt-card pt-elevation-2 smd-pointer ${userClasseName}`} key={index} onClick={(e) => {
                this.setState({ selected: index });
                this.onUserClick(user, addresses, index);
              }}>
                {user}
              </div>
            </div>
          </div>
        );
      }.bind(this));

    return (
      <div>
        <div className="row">
          <div className="col-sm-4 text-left">
            <h3>Accounts</h3>
          </div>
          <div className="col-sm-8 text-right">
            <div className="pt-button-group">
              <SendTokens />
              <CreateBlocUser />
            </div>
          </div>
        </div>
        <div className="row">
          <div className="col-sm-4">
            <div className="pt-input-group pt-dark pt-large">
              <span className="pt-icon pt-icon-search"></span>
              <input
                className="pt-input"
                type="search"
                placeholder="Search accounts"
                onChange={e => this.updateFilter(e.target.value.toLowerCase())}
                dir="auto" />
            </div>
          </div>
        </div>
        <div className="container-fluid pt-dark">
          <div className="row">
            <div className="col-sm-4 main-div">
              <div className="accounts-margin-top">
                {!isModeOauth && (rows.length === 0
                  ?
                  <table>
                    <tbody>
                      <tr>
                        <td colSpan={3}>No Accounts</td>
                      </tr>
                    </tbody>
                  </table>
                  : rows)}
                {isModeOauth && this.renderOauthAccounts()}
              </div>
            </div>
            <div className="col-sm-8 account-details">
              <div>
                {selectedAddresses.length ? selectedAddresses : null}
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export function mapStateToProps(state) {
  return {
    accounts: state.accounts.accounts,
    filter: state.accounts.filter,
    selectedChain: state.chains.selectedChain
  };
}

export default withRouter(
  connect(mapStateToProps,
    {
      fetchAccountDetail,
      fetchUserAddresses,
      fetchAccounts,
      changeAccountFilter,
      resetUserAddress,
      fetchOauthAccounts,
    }
  )(BlocAccounts));
