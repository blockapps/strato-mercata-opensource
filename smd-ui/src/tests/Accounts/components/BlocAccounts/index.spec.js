import React from 'react';
import BlocAccounts, { mapStateToProps } from '../../../../components/Accounts/components/BlocAccounts';
import { indexAccountsMock, filter } from '../../accountsMock';

describe('BlocAccounts: index', () => {

  describe('render with', () => {
    test('empty values', () => {
      const props = {
        accounts: [],
        filter: '',
        history: {},
        fetchAccounts: jest.fn(),
        changeAccountFilter: jest.fn(),
        faucetRequest: jest.fn(),
        resetUserAddress: jest.fn(),
        fetchUserAddresses: jest.fn()
      }
      const wrapper = shallow(
        <BlocAccounts.WrappedComponent {...props} />
      );

      expect(wrapper.debug()).toMatchSnapshot();
    });

    test('mocked values', () => {
      const props = {
        accounts: indexAccountsMock,
        filter: '',
        history: {},
        fetchAccounts: jest.fn(),
        changeAccountFilter: jest.fn(),
        faucetRequest: jest.fn(),
        resetUserAddress: jest.fn(),
        fetchUserAddresses: jest.fn()
      }
      const wrapper = shallow(
        <BlocAccounts.WrappedComponent {...props} />
      );

      expect(wrapper.debug()).toMatchSnapshot();
    });

  });

  test('hide account on click', () => {
    const props = {
      accounts: indexAccountsMock,
      filter: '',
      history: {},
      fetchAccounts: jest.fn(),
      changeAccountFilter: jest.fn(),
      faucetRequest: jest.fn(),
      resetUserAddress: jest.fn(),
      fetchUserAddresses: jest.fn()
    };

    const wrapper = shallow(
      <BlocAccounts.WrappedComponent {...props} />
    );

    wrapper.find('div').at(14).simulate('click');
    expect(props.resetUserAddress).toHaveBeenCalled();
    expect(props.resetUserAddress).toHaveBeenCalledTimes(1);
    expect(props.fetchUserAddresses).not.toHaveBeenCalled();
  });

  test('open account on click', () => {
    const props = {
      accounts: indexAccountsMock,
      filter: '',
      history: { push: jest.fn() },
      fetchAccounts: jest.fn(),
      changeAccountFilter: jest.fn(),
      faucetRequest: jest.fn(),
      resetUserAddress: jest.fn(),
      fetchUserAddresses: jest.fn()
    }
    const wrapper = shallow(
      <BlocAccounts.WrappedComponent {...props} />
    );

    wrapper.find('div').at(59).simulate('click');
    expect(props.fetchUserAddresses).toHaveBeenCalled();
    expect(props.fetchUserAddresses).toHaveBeenCalledTimes(1);
    expect(props.resetUserAddress).not.toHaveBeenCalled();
  });

  test('invoke onchange on input and trigger changeAccountFilter', () => {
    const props = {
      accounts: indexAccountsMock,
      filter: '',
      history: { push: jest.fn() },
      fetchAccounts: jest.fn(),
      changeAccountFilter: jest.fn(),
      faucetRequest: jest.fn(),
      resetUserAddress: jest.fn(),
      fetchUserAddresses: jest.fn()
    }
    const wrapper = shallow(
      <BlocAccounts.WrappedComponent {...props} />
    );

    wrapper.find('input').simulate('change', { target: { name: "pollName", value: "spam" } });
    expect(props.changeAccountFilter).toHaveBeenCalled();
  });

  test('mapStateToProps with default values', () => {
    const state = {
      accounts: {
        faucet: {
          address: 'd2263b71c14010ff03d8f786670aba691b22b158',
          status: false,
        },
        accounts: indexAccountsMock,
        filter: filter,
      },
      chains: {
        selectedChain: 'airline cartel 1'
      }
    }
    expect(mapStateToProps(state)).toMatchSnapshot();
  });

});