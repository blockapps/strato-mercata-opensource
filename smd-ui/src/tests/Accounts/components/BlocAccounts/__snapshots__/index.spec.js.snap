// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlocAccounts: index mapStateToProps with default values 1`] = `
Object {
  "accounts": Object {
    "123456789012345678901234567890123": Object {
      "2327839d6c7fc1188085554193cf578537ee660d": Object {
        "error": null,
      },
    },
    "Admin_1085_64667": Object {
      "d2263b71c14010ff03d8f786670aba691b22b158": Object {
        "error": null,
      },
    },
    "Admin_1177_31203": Object {
      "d46ade1b4fb6360fad2d05efd5d6d9fb51849320": Object {
        "error": null,
      },
    },
    "Admin_1177_39206": Object {
      "b80c7ad40cd178479d480e7136d93f2302692e4d": Object {
        "error": null,
      },
    },
    "Admin_1177_49507": Object {
      "0bdd9ade6477ba753650cc5d9ce40a17c42246c1": Object {
        "error": null,
      },
    },
    "Admin_1301_62816": Object {
      "45b53a8b688c3a3faca30e5842800922f80fdb8d": Object {
        "error": null,
      },
      "60122a032bebcb89e228899b60c4d190b5a17aa3": Object {
        "error": null,
      },
    },
    "Admin_1301_86023": Object {
      "585b8e9f36f43f38d9118b8750a5e6cd753e382e": Object {
        "error": null,
      },
    },
    "Buyer1": Object {
      "044eda43ba9c76fc36b9183c96f7a8fad8d21fe6": Object {
        "error": null,
      },
    },
    "Buyer1177_18955": Object {
      "73ea1c503b915e28162374591670904732d0e729": Object {
        "error": null,
      },
    },
    "Buyer2": Object {
      "ae5bfd0fc750d1d99e28e733bdac933dac826a4f": Object {
        "error": null,
      },
    },
    "Supplier0_1301_46441": Object {
      "error": "error",
    },
    "Supplier1": Object {
      "370adf114257cb0e0025eedf0a96261b51af23e3": Object {
        "error": null,
      },
    },
    "Supplier1177_62701": Object {
      "21edfc77121d05aba26955b6cd435bd8f864ad78": Object {
        "error": null,
      },
    },
    "Supplier1_1301_46441": Object {
      "0ca4a9bb1b37e6d63d142722e7c37bbf8f959547": Object {
        "error": null,
      },
    },
    "Supplier2": Object {
      "bc7199468816b3b92e54828affa08a010e9adb02": Object {
        "error": null,
      },
    },
    "Supplier2_1301_46441": Object {
      "57f2ed9058d2e868d7bbae9db03af12d27675117": Object {
        "error": null,
      },
    },
    "User_1177_14462": Object {
      "cd0046de8051af7e2781b58e892b25467133276f": Object {
        "error": null,
      },
    },
    "User_1177_26292": Object {
      "dcbdac4d738461fcf11669b73668211371839c76": Object {
        "error": null,
      },
    },
    "User_1177_37717": Object {
      "9677e59a2de0a2e2f712e1d76720857066a396b9": Object {
        "error": null,
      },
    },
    "User_1177_42489": Object {
      "a421b5e3c34118266745dd8426a80b35513b2277": Object {
        "error": null,
      },
    },
    "User_1177_52307": Object {
      "0e29350e6de10d60e58ffc2a7ad84401af5f2f29": Object {
        "error": null,
      },
    },
    "User_1177_59359": Object {
      "cf7a807b9bbb64cab5f988bddb4bedfd066a83de": Object {
        "error": null,
      },
    },
    "User_1177_59578": Object {
      "620083a3803fc8a5ad71e403c1a72a0e73ece645": Object {
        "error": null,
      },
      "dcbfca1d9a80d2d78fa6f74d62489a8ff4bd9c5f": Object {
        "error": null,
      },
    },
    "User_1177_69645 ?#%!@*": Object {},
    "User_1177_79118": Object {
      "33c02a81e677ea493aace735f0a6b44cfa18f44e": Object {
        "error": null,
      },
    },
  },
  "filter": "Search accounts",
  "selectedChain": "airline cartel 1",
}
`;

exports[`BlocAccounts: index render with empty values 1`] = `
"<div>
  <div className=\\"row\\">
    <div className=\\"col-sm-4 text-left\\">
      <h3>
        Accounts
      </h3>
    </div>
    <div className=\\"col-sm-8 text-right\\">
      <div className=\\"pt-button-group\\">
        <withRouter(Connect(ReduxForm)) />
        <withRouter(Connect(ReduxForm)) />
      </div>
    </div>
  </div>
  <div className=\\"row\\">
    <div className=\\"col-sm-4\\">
      <div className=\\"pt-input-group pt-dark pt-large\\">
        <span className=\\"pt-icon pt-icon-search\\" />
        <input className=\\"pt-input\\" type=\\"search\\" placeholder=\\"Search accounts\\" onChange={[Function: onChange]} dir=\\"auto\\" />
      </div>
    </div>
  </div>
  <div className=\\"container-fluid pt-dark\\">
    <div className=\\"row\\">
      <div className=\\"col-sm-4 main-div\\">
        <div className=\\"accounts-margin-top\\">
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                length
              </div>
            </div>
          </div>
        </div>
      </div>
      <div className=\\"col-sm-8 account-details\\">
        <div />
      </div>
    </div>
  </div>
</div>"
`;

exports[`BlocAccounts: index render with mocked values 1`] = `
"<div>
  <div className=\\"row\\">
    <div className=\\"col-sm-4 text-left\\">
      <h3>
        Accounts
      </h3>
    </div>
    <div className=\\"col-sm-8 text-right\\">
      <div className=\\"pt-button-group\\">
        <withRouter(Connect(ReduxForm)) />
        <withRouter(Connect(ReduxForm)) />
      </div>
    </div>
  </div>
  <div className=\\"row\\">
    <div className=\\"col-sm-4\\">
      <div className=\\"pt-input-group pt-dark pt-large\\">
        <span className=\\"pt-icon pt-icon-search\\" />
        <input className=\\"pt-input\\" type=\\"search\\" placeholder=\\"Search accounts\\" onChange={[Function: onChange]} dir=\\"auto\\" />
      </div>
    </div>
  </div>
  <div className=\\"container-fluid pt-dark\\">
    <div className=\\"row\\">
      <div className=\\"col-sm-4 main-div\\">
        <div className=\\"accounts-margin-top\\">
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer  selected\\" onClick={[Function: onClick]}>
                Admin_1085_64667
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Supplier1
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Supplier2
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Buyer1
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Buyer2
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Admin_1177_49507
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Admin_1177_39206
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Admin_1177_31203
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                User_1177_14462
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                123456789012345678901234567890123
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                User_1177_26292
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                User_1177_69645 ?#%!@*
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                User_1177_59578
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                User_1177_79118
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                User_1177_59359
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                User_1177_42489
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                User_1177_37717
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                User_1177_52307
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Buyer1177_18955
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Supplier1177_62701
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Admin_1301_86023
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Admin_1301_62816
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Supplier0_1301_46441
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Supplier1_1301_46441
              </div>
            </div>
          </div>
          <div className=\\"smd-margin-8\\">
            <div className=\\"row\\">
              <div className=\\"pt-card pt-elevation-2 smd-pointer \\" onClick={[Function: onClick]}>
                Supplier2_1301_46441
              </div>
            </div>
          </div>
        </div>
      </div>
      <div className=\\"col-sm-8 account-details\\">
        <div>
          <withRouter(Connect(Account)) name=\\"Admin_1085_64667\\" address=\\"d2263b71c14010ff03d8f786670aba691b22b158\\" />
        </div>
      </div>
    </div>
  </div>
</div>"
`;
